### ISSUE TITLE
["BUG"] Requests but doesn't use Stream mode for AI tools

### ISSUE THREAD

**Issue by iandol** [x-cmd/x-cmd#42]
**Describe the bug**

You are still sending `"stream": true` to OpenAI compatible services (I use LM Studio, in this case I configured OpenAI to use a local endpoint) but I don't think you support streaming responses? Here is the request that gets sent (using `x writer --sendalias zh` command):

```
[2024-07-09 10:07:38.905] [INFO] Received POST request to /v1/chat/completions with body: {
  "model": "gpt-4o",
  "messages": [
    {
      "role": "system",
      "content": "\nYou are a translate assistant to help the following text to 'Chinese' .\nNOTICE: Your response MUST ONLY be the text translated. NO extra.\n\nThe text to be translated is as below:\n"
    },
    {
      "role": "user",
      "content": "What do I need?"
    }
  ],
  "max_tokens": 1000,
  "seed": 0,
  "stream": true
}
```

Ideally it would be nice to support streaming responses, but in the short term just send `"stream": false`?

**System (please complete the following information):**
 - Version: x7 f0e0c03001b74e
 - Shell: Elvish
 - OS: Darwin
 - Arch: arm64

**Comment by edwinjhlee** [comment]
> **Describe the bug**
> 
> You are still sending `"stream": true` to OpenAI compatible services (I use LM Studio, in this case I configured OpenAI to use a local endpoint) but I don't think you support streaming responses? Here is the request that gets sent (using `x writer --sendalias zh` command):
> 
> ```
> [2024-07-09 10:07:38.905] [INFO] Received POST request to /v1/chat/completions with body: {
>   "model": "gpt-4o",
>   "messages": [
>     {
>       "role": "system",
>       "content": "\nYou are a translate assistant to help the following text to 'Chinese' .\nNOTICE: Your response MUST ONLY be the text translated. NO extra.\n\nThe text to be translated is as below:\n"
>     },
>     {
>       "role": "user",
>       "content": "What do I need?"
>     }
>   ],
>   "max_tokens": 1000,
>   "seed": 0,
>   "stream": true
> }
> ```
> 
> Ideally it would be nice to support streaming responses, but in the short term just send `"stream": false`?
> 
> **System (please complete the following information):**
> 
> * Version: x7 f0e0c03001b74e
> * Shell: Elvish
> * OS: Darwin
> * Arch: arm64


This is history issue. We actually supports stream mode 1 year ago.

However, the mardown rendering is not good at stream mode. And also, eventually, the model will response faster and faster. So stream mode will become useless in the future.

So we decided to just render the data after data is received, since 4 months ago.

---

However, recently I just figure out a new way to leverage it, and we just try to implement it in recent versions.

1. We still render the data as a whole document.
2. However, during the streaming, we will previde a rotate view of received markdown, which will greatly improve userability.


Still there is a lot of work to us. Because there are openai,ollama,kimi,lms,mistral,gemini ... Our chat module is designed as an unified CLI for all of these providers. We have to ensure all providers works and behave in consistent UI.


Also, it is quite frustating that only openai didn't return the token used during stream mode. I am now curious whether openai provide this information.

**Comment by iandol** [comment]
Right, so many APIs it's annoying. At least for OpenAI compatible tools (which many are), you can still turn off stream mode unitl you have a stream UI. Streaming causes more network traffic and so may be slower if you are not using it anyway...

**Comment by edwinjhlee** [comment]
We implemented a rotating view while receiving data from the streaming API to enhance usability. The markdown data is then rendered as a complete document upon completion.
    
Please confirm if this resolves the issue.

**Comment by iandol** [comment]
Yes, this approach works well, although I'll need a different colour theme:

<img width="950" alt="image" src="https://github.com/user-attachments/assets/8960c2de-8347-4a99-8c60-ab11847e8517">

One remaining issue is the first token is lost, note the first part of the replies:

<img width="466" alt="image" src="https://github.com/user-attachments/assets/87a78042-2ab3-479d-8c7c-b19cfe195f12">

"The " is missing in first reposnse and "Vol" in the second resposne

**Comment by edwinjhlee** [comment]
OK.

**Comment by edwinjhlee** [comment]
> Yes, this approach works well, although I'll need a different colour theme:
> 
> <img alt="image" width="950" src="https://private-user-images.githubusercontent.com/153905/353273063-8960c2de-8347-4a99-8c60-ab11847e8517.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ0ODYxNDQsIm5iZiI6MTcyNDQ4NTg0NCwicGF0aCI6Ii8xNTM5MDUvMzUzMjczMDYzLTg5NjBjMmRlLTgzNDctNGE5OS04YzYwLWFiMTE4NDdlODUxNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgyNFQwNzUwNDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iODYyYmYxNGM5OTMzZjVmMjExMDgzYmEwZDlmMzVlMjBhN2U5NjVhMGFkNWVjOWZmNzg5OWU5MTA5OGViM2NiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Z1oX1vMSYMg14nzoRJ0m8SmDbKymuPEyuKVCpictBeA">
> One remaining issue is the first token is lost, note the first part of the replies:
> 
> <img alt="image" width="466" src="https://private-user-images.githubusercontent.com/153905/353273201-87a78042-2ab3-479d-8c7c-b19cfe195f12.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ0ODYxNDQsIm5iZiI6MTcyNDQ4NTg0NCwicGF0aCI6Ii8xNTM5MDUvMzUzMjczMjAxLTg3YTc4MDQyLTJhYjMtNDc5ZC04YzdjLWIxOWNmZTE5NWYxMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgyNFQwNzUwNDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03ZjAxZTQ1YjI1MzJhN2ZmZmQxNTU0N2VjOTQ1OTM2NGMxMGNiNmZiMDYyYzk4YTZiYjEzYzI5NzlmMzc3OTY2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Ns34R_mAxXIpDdLhVamVPbN-y-VmnHN8ibtInYTV7Qo">
> "The " is missing in first reposnse and "Vol" in the second resposne


I am consdiering using a better foreground color in v0.4.8 ~
